### PR TITLE
Range operations performance

### LINKED ISSUE #103: Range operations on bitmap containers can be inefficient
A lot of times, we call for a full computation of the bitmap container's cardinality even when it is inefficient to do so. This affects methods such as addRange, removeRange, not, and all operations between a bitmap container and a run container.

### PR DESCRIPTION
**Pull request by ppiotrow**
Fixes #103

### CODE REVIEW

**Comment by lemire** [comment]
Excellent. I think the current status is as follows...

- [X] MappeableBitmapContainer.add(int begin, int end)
- [ ] MappeableBitmapContainer.iadd(int begin, int end)
- [ ] MappeableBitmapContainer.remove(int begin, int end)
- [ ] MappeableBitmapContainer.iremove(int begin, int end)

**Comment by lemire** [comment]
@ppiotrow 

I think that the current status is as follows...


- [X] BitmapContainer.add(int begin, int end)
- [X] BitmapContainer.iadd(int begin, int end)
- [ ] BitmapContainer.remove(int begin, int end)
- [ ] BitmapContainer.iremove(int begin, int end)
- [ ] MappeableBitmapContainer.add(int begin, int end)
- [ ] MappeableBitmapContainer.iadd(int begin, int end)
- [ ] MappeableBitmapContainer.remove(int begin, int end)
- [ ] MappeableBitmapContainer.iremove(int begin, int end)

**Comment by ppiotrow** [comment]
Last two commits added optimization in `ior` and `or` between Bitmap and Run Containers as you suggested in issue description.
>  and all operations between a bitmap container and a run container.

File: src/main/java/org/roaringbitmap/RunContainer.java [hunk 1]
```diff
@@ -609,12 +609,16 @@ public Container andNot(BitmapContainer x) {
...
+      int flippedOnes = Util.cardinalityInBitmapRange(answer.bitmap, start, end);
       Util.resetBitmapRange(answer.bitmap, lastPos, start);
       Util.flipBitmapRange(answer.bitmap, start, end);
+      answer.updateCardinality(prevOnes + flippedOnes, end - start - flippedOnes);
       lastPos = end;
     }
-    Util.resetBitmapRange(answer.bitmap, lastPos, answer.bitmap.length * 64);
-    answer.computeCardinality();
+    int prevOnes = Util.cardinalityInBitmapRange(answer.bitmap, lastPos, BitmapContainer.MAX_CAPACITY);
+    Util.resetBitmapRange(answer.bitmap, lastPos, BitmapContainer.MAX_CAPACITY);
```
Comment (ppiotrow) [inline src/main/java/org/roaringbitmap/RunContainer.java]: I've replaced ` answer.bitmap.length * 64` with `BitmapContainer.MAX_CAPACITY`. Good/bad?

Comment (lemire) [inline src/main/java/org/roaringbitmap/RunContainer.java]: At a glance, it looks fine.

**Comment by lemire** [comment]
Make sure to pass our style requirements. (It is annoying, but it keeps formatting issues under control.)

**Comment by lemire** [comment]
@ppiotrow 

To preserve symmetry, I recommend you port over your work to the ``buffer`` package (e.g., ``MappeableBitmapContainer`` and so forth). This is slightly annoying, but it is often mostly just copy-paste with straightforward adaptations. (Some key users of the library rely on the ``buffer`` package and we don't want the code to diverge as it would make maintainability increasingly more difficult.)

**Comment by ppiotrow** [comment]
I think I'm done with non-buffer containers. Next commit will be on buffered versions.

I'd like to mark `Util.cardinalityInBitmapWordRange`, `Util.(set|flip|reset)BitmapRangeAndCardinalityChange` with @Deprecated, what do you think?

**Comment by lemire** [comment]
> I'd like to mark Util.cardinalityInBitmapWordRange, Util.(set|flip|reset)BitmapRangeAndCardinalityChange with @Deprecated, what do you think?

Yes. That sounds reasonable.

**Comment by ppiotrow** [comment]
Yes, thank you. Very good piece of CI ;)

**Comment by ppiotrow** [comment]
Optimization applied wherever possible. The usages of `computeCardinality` reduced only to repairing  after lazy operations.
Right now I'm looking at the code coverage reports and it seems that few more tests need to be added.
Second thing is that I'd like to benchmark those optimizations before merge.

**Comment by ppiotrow** [comment]
I need help answering my question.
The `testRangeCardinality4` [test](https://github.com/RoaringBitmap/RoaringBitmap/pull/133/files#diff-a3129b26395ab93fa5ed94038525e9d1R277) suposed to check the method `andNot` version NOT backed by simple array.  I was surprised that is has not happened [report](https://coveralls.io/builds/9566665/source?filename=src%2Fmain%2Fjava%2Forg%2Froaringbitmap%2Fbuffer%2FMappeableBitmapContainer.java#L326). It appeared the execution of `this.clone()` at the beginning of the method in my case ALWAYS creates an instance "backed by simple array". This means that the `else` block is a dead code. I'd say it's a bug or if it is expected behavior is should be explicitly visible in code.

**Comment by lemire** [comment]
@ppiotrow Agreed, this is dead code.

**Comment by ppiotrow** [comment]
This is out of this issue's scope and I've created new one #137 . I'll focus only on coverage of the code that is not dead.

**Comment by lemire** [comment]
Agreed.

**Comment by ppiotrow** [comment]
```
===BEFORE=============
Benchmark                                   Before                 After
RangeOperationBenchmark.computeCardinality  2568103,459 ops/s      2643159,521 ops/s
RangeOperationBenchmark.updateCardinality   X                      366903155,877 ops/s
RangeOperationBenchmark.iadd                2608901,656 ops/s      60581004,868 ops/s
RangeOperationBenchmark.iand                1929469,692 ops/s      1936434,587 ops/s
RangeOperationBenchmark.iandNot             2323081,302 ops/s      10376728,710 ops/s
RangeOperationBenchmark.inot                51287863,801 ops/s     59930300,677 ops/s
RangeOperationBenchmark.inotBig             2094809,672 ops/s      2662481,679 ops/s
RangeOperationBenchmark.inotFull            7120530,040 ops/s      1960495,751 ops/s
RangeOperationBenchmark.ior                 2332982,167 ops/s      10348721,207 ops/s
RangeOperationBenchmark.iremove             2598827,878 ops/s      70233643,894 ops/s
RangeOperationBenchmark.ixor                2279555,274 ops/s      9398872,176 ops/s
```

The code I was comparing my changes was obtained by:
git checkout aebc876f99523db452c7ec47f0f415a9508b9b9a -- src/main/java/org/roaringbitmap/BitmapContainer.java
git checkout aebc876f99523db452c7ec47f0f415a9508b9b9a -- src/main/java/org/roaringbitmap/RunContainer.java

Assumptions:
Whenever having choice to benchmark `iMethod()` vs. `method()` i always choose i-version to not benchmark allocations.
Benchmark data chosen to avoid execution of `toArrayContainer()`

My comments:
In general I can see great improvement in operations per second. The results definitely depends on the layout of `RunContainer`: number of runs, run lengths. I think there can be done more to have improvements in `not()` benchmarks.

**Comment by lemire** [comment]
Impressive!

Do you know what the issue is with respect to ``RangeOperationBenchmark.inotFull ``?

c.c. @gssiyankai

**Comment by ppiotrow** [comment]
The issue with full negation is that previous optimization was better there 
`cardinality = MAX_CAPACITY - cardinality`
I'll get this optimization back and try to improve other things using benchmarks with next commit.

**Comment by ppiotrow** [comment]
The new optimization is "copied" from previous `inot` method. If the range `<start, end)` to count cardinality is bigger than half of maximum range, then the cardinality of that range is equal to `cardinality` - `cardinalityInRange(0,start)` - `cardinalityInRange(end, MAX_CARDINALITY)`.

Benchmarks with significant change after new optimization.
```
Benchmark                                   Before                 After           NewOptimization
1)RangeOperationBenchmark.iadd                2 608 901 ops/s     60 581 004 ops/s  74 556 755
2)RangeOperationBenchmark.iand                1 929 469 ops/s      1 936 434 ops/s   4 595 676
3)RangeOperationBenchmark.inotBig             2 094 809 ops/s      2 662 481 ops/s   4 664 188
4)RangeOperationBenchmark.inotFull            7 120 530 ops/s      1 960 495 ops/s   7 246 319
```
Explanation of improvement.
1) No idea what caused improvement, probably benchmark environment is not stable.
2) Last operation in `iand` is huge `cardinalityInRange(start, MAX_CAPACITY);`
3) There was huge run container in benchmark data perfect for the new optimization.
4) The new optimization actually reverts previous optimization from `init` but applies it to other `BitmapContainer` methods.

**Comment by lemire** [comment]
@ppiotrow 

Fantastic.

I can give you access to the test server I use (Linux). It does not guarantee reproducible results, but it might help.

File: src/main/java/org/roaringbitmap/RunContainer.java [hunk 2]
```diff
@@ -357,11 +357,14 @@ public Container and(BitmapContainer x) {
...
       int end = Util.toIntUnsigned(this.getValue(rlepos));
+      int prevOnes = Util.cardinalityInBitmapRange(answer.bitmap, start, end);
       Util.resetBitmapRange(answer.bitmap, start, end); // had been x.bitmap
+      answer.updateCardinality(prevOnes, 0);
       start = end + Util.toIntUnsigned(this.getLength(rlepos)) + 1;
     }
-    Util.resetBitmapRange(answer.bitmap, start, Util.maxLowBitAsInteger() + 1); // had been x.bitmap
-    answer.computeCardinality();
+    int ones = Util.cardinalityInBitmapRange(answer.bitmap, start, BitmapContainer.MAX_CAPACITY);
+    Util.resetBitmapRange(answer.bitmap, start, BitmapContainer.MAX_CAPACITY); // had been x.bitmap
```
Comment (lemire) [inline src/main/java/org/roaringbitmap/RunContainer.java]: Yes. I think you replaced 0xFFFF + 1 with 1<<16, which is fine.

**Comment by lemire** [comment]
Ok. This looks like fantastic work. Should we merge this and try to make a release?

c.c. @gssiyankai

**Comment by gssiyankai** [comment]
Good work. Let's merge.

**Comment by ppiotrow** [comment]
@lemire In my opinion it is ready for merge&release. Regarding the benchmark linux server I can try that (even if it would be post-merge benchmark run).

**Comment by ppiotrow** [comment]
Maybe last question: how to be sure that whenever execution steps into the new optimizations the BitmapContainer is not in unrepaired (cardinality == -1) state?

**Comment by lemire** [comment]
@ppiotrow 

Email me privately at lemire@gmail.com if you are interested in getting access to a Linux test box for the future. I'll give you access.

**Comment by lemire** [comment]
> Maybe last question: how to be sure that whenever execution steps into the new optimizations the BitmapContainer is not in unrepaired (cardinality == -1) state?

That should never happen in the normal course of things, unless someone mutate the bitmaps from various threads (and then they are lost anyhow).

**Comment by lemire** [comment]
Merged and release.

**Comment by lemire** [comment]
Version 0.6.34.

**Pull request merged**